File: SecretFinderDebug.js
// SecretFinder.js
// Assignment solution in JavaScript (Node.js)

function parseShare(x, base, value) {
    return { x: parseInt(x), y: parseInt(value, base) };
}

// Lagrange interpolation at x = 0
function lagrangeInterpolation(shares) {
    let secret = 0;

    for (let i = 0; i < shares.length; i++) {
        let xi = shares[i].x;
        let yi = shares[i].y;

        let term = yi;
        for (let j = 0; j < shares.length; j++) {
            if (i !== j) {
                let xj = shares[j].x;
                term *= (0 - xj) / (xi - xj);
            }
        }
        secret += term;
    }
    return Math.round(secret);
}

// Main function to solve assignment
function findSecret(jsonInput) {
    const n = jsonInput.keys.n;
    const k = jsonInput.keys.k;

    // Step 1: decode all shares
    let shares = [];
    for (let key in jsonInput) {
        if (key !== "keys") {
            shares.push(parseShare(key, jsonInput[key].base, jsonInput[key].value));
        }
    }

    // Step 2: generate all subsets of size k
    function getCombinations(arr, k) {
        if (k === 0) return [[]];
        if (arr.length === 0) return [];
        let [first, ...rest] = arr;

        let withFirst = getCombinations(rest, k - 1).map(c => [first, ...c]);
        let withoutFirst = getCombinations(rest, k);

        return withFirst.concat(withoutFirst);
    }

    let subsets = getCombinations(shares, k);

    // Step 3: calculate secrets for each subset
    let results = {};
    for (let subset of subsets) {
        let secret = lagrangeInterpolation(subset);
        results[secret] = (results[secret] || 0) + 1;
    }

    // Step 4: find majority secret
    let secret = Object.entries(results).reduce((a, b) => (a[1] > b[1] ? a : b))[0];

    return { secret: parseInt(secret), allResults: results };
}

// ----------------------------
// Run on sample test case
// ----------------------------
const sample = {
    "keys": { "n": 4, "k": 3 },
    "1": { "base": "10", "value": "4" },
    "2": { "base": "2", "value": "111" },
    "3": { "base": "10", "value": "12" },
    "6": { "base": "4", "value": "213" }
};

const result = findSecret(sample);
console.log("Final Secret:", result.secret);
console.log("All Subset Results:", result.allResults);

File: SecretFinderDebug.java
import org.json.JSONObject;
import java.math.BigInteger;
import java.util.*;

public class SecretFinderDebug {

    static BigInteger toDecimal(String value, int base) {
        return new BigInteger(value, base);
    }

    static BigInteger lagrangeAtZero(List<int[]> subset) {
        BigInteger result = BigInteger.ZERO;

        for (int i = 0; i < subset.size(); i++) {
            int xi = subset.get(i)[0];
            BigInteger yi = BigInteger.valueOf(subset.get(i)[1]);

            double numerator = 1, denominator = 1;
            for (int j = 0; j < subset.size(); j++) {
                if (i != j) {
                    int xj = subset.get(j)[0];
                    numerator *= (0 - xj);
                    denominator *= (xi - xj);
                }
            }

            double term = yi.doubleValue() * (numerator / denominator);
            result = result.add(BigInteger.valueOf(Math.round(term)));
        }

        return result;
    }

    static List<List<int[]>> generateSubsets(List<int[]> shares, int k) {
        List<List<int[]>> subsets = new ArrayList<>();
        generate(shares, k, 0, new ArrayList<>(), subsets);
        return subsets;
    }

    static void generate(List<int[]> shares, int k, int index, List<int[]> current, List<List<int[]>> subsets) {
        if (current.size() == k) {
            subsets.add(new ArrayList<>(current));
            return;
        }
        if (index >= shares.size()) return;

        current.add(shares.get(index));
        generate(shares, k, index + 1, current, subsets);
        current.remove(current.size() - 1);
        generate(shares, k, index + 1, current, subsets);
    }

    public static void main(String[] args) {
        String jsonInput = """
        {
          "keys": { "n": 4, "k": 3 },
          "1": { "base": "10", "value": "4" },
          "2": { "base": "2",  "value": "111" },
          "3": { "base": "10", "value": "12" },
          "6": { "base": "4",  "value": "213" }
        }
        """;

        JSONObject obj = new JSONObject(jsonInput);
        JSONObject keys = obj.getJSONObject("keys");
        int n = keys.getInt("n");
        int k = keys.getInt("k");

        // Decode shares
        List<int[]> shares = new ArrayList<>();
        System.out.println("🔎 Decoding shares:");
        for (String key : obj.keySet()) {
            if (!key.equals("keys")) {
                int x = Integer.parseInt(key);
                JSONObject entry = obj.getJSONObject(key);
                int base = Integer.parseInt(entry.getString("base"));
                String value = entry.getString("value");
                BigInteger y = toDecimal(value, base);
                shares.add(new int[]{x, y.intValue()});
                System.out.println("  (x=" + x + ", y=" + y + ")");
            }
        }

        // Generate subsets and calculate secrets
        System.out.println("\n🔎 Evaluating subsets of size " + k + ":");
        List<List<int[]>> subsets = generateSubsets(shares, k);
        Map<BigInteger, Integer> frequency = new HashMap<>();

        for (List<int[]> subset : subsets) {
            BigInteger secret = lagrangeAtZero(subset);
            frequency.put(secret, frequency.getOrDefault(secret, 0) + 1);

            System.out.print("  Subset { ");
            for (int[] s : subset) {
                System.out.print("(" + s[0] + "," + s[1] + ") ");
            }
            System.out.println("} → Secret = " + secret);
        }

        // Majority vote
        BigInteger correctSecret = Collections.max(frequency.entrySet(), Map.Entry.comparingByValue()).getKey();
        System.out.println("\n✅ Correct Secret (majority vote) = " + correctSecret);

        // Detect wrong shares
        System.out.println("\n⚠️ Checking for wrong shares...");
        Set<Integer> wrongShares = new HashSet<>();
        for (int[] share : shares) {
            List<int[]> temp = new ArrayList<>(shares);
            temp.remove(share);

            List<List<int[]>> subSubsets = generateSubsets(temp, k);
            Map<BigInteger, Integer> freqCheck = new HashMap<>();

            for (List<int[]> sub : subSubsets) {
                BigInteger sec = lagrangeAtZero(sub);
                freqCheck.put(sec, freqCheck.getOrDefault(sec, 0) + 1);
            }

            BigInteger mostCommon = Collections.max(freqCheck.entrySet(), Map.Entry.comparingByValue()).getKey();
            if (!mostCommon.equals(correctSecret)) {
                wrongShares.add(share[0]);
            }
        }

        System.out.println("❌ Wrong Shares = " + wrongShares);
    }
}
